File: ui/src/commands/player.ts
// Copyright (C) Tsukilabs contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from '@tauri-apps/api/core';

export function getPlayer(id: PlayerId) {
  return invoke<Player>('get_player', { id });
}

export async function getPlayerCoords(id: PlayerId) {
  return invoke<readonly Coord[]>('get_player_coords', { id });
}

export function getPlayers() {
  return invoke<readonly Player[]>('get_players');
}

export function playerExists(id: PlayerId) {
  return invoke<boolean>('player_exists', { id });
}

export function setPlayerStatus(id: PlayerId, status: PlayerStatus) {
  return invoke<null>('set_player_status', { id, status });
}

export function spawnPlayer(options: PlayerOptions) {
  return invoke<null>('spawn_player', { options });
}

export function spawnPlayerVillage(id: PlayerId) {
  return invoke<null>('spawn_player_village', { id });
}

File: crates/nil/src/command/player.rs
// Copyright (C) Tsukilabs contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::error::Result;
use crate::manager::ManagerExt;
use nil_core::player::{Player, PlayerId, PlayerOptions, PlayerStatus};
use nil_core::village::Coord;
use tauri::AppHandle;

#[tauri::command]
pub async fn get_player(app: AppHandle, id: PlayerId) -> Result<Player> {
  app
    .client(async |cl| cl.get_player(id).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_player_coords(app: AppHandle, id: PlayerId) -> Result<Vec<Coord>> {
  app
    .client(async |cl| cl.get_player_coords(id).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn get_players(app: AppHandle) -> Result<Vec<Player>> {
  app
    .client(async |cl| cl.get_players().await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn player_exists(app: AppHandle, id: PlayerId) -> Result<bool> {
  app
    .client(async |cl| cl.player_exists(id).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn set_player_status(app: AppHandle, id: PlayerId, status: PlayerStatus) -> Result<()> {
  app
    .client(async |cl| cl.set_player_status(id, status).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn spawn_player(app: AppHandle, options: PlayerOptions) -> Result<()> {
  app
    .client(async |cl| cl.spawn_player(options).await)
    .await?
    .map_err(Into::into)
}

#[tauri::command]
pub async fn spawn_player_village(app: AppHandle, id: PlayerId) -> Result<()> {
  app
    .client(async |cl| cl.spawn_player_village(id).await)
    .await?
    .map_err(Into::into)
}
